### ISSUE TITLE
Add support for Mistral3 and Ministral models

### ISSUE THREAD

**Issue by arielge** [torch-spyre/hf-adapters#76]
Current status:

- [x] Text-only Mistral3 (#83)
- [x] Text-only Ministral3 (#103)
- [x] Full multimodal generation for Mistral3 (#127)
- [x] Full multimodal generation for Ministral3 (#176)

**Comment by 35Upesh** [comment]
**Implementation Repository:** I forked the `hf-adapters` repository and implemented all Ministral-3 onboarding changes in my fork. The complete code changes and related updates can be found here in branch `ministral`:
 https://github.com/35Upesh/hf-adapters.git

# Ministral-3 14B Instruct Onboarding – Incorrect Output Generation

## Model Information
- **Model ID**: `mistralai/Ministral-3-14B-Instruct-2512`
- **Model Type**: `mistral3` (multimodal config with nested text decoder)
- **Architecture**: Standard GQA decoder, no sliding window attention
- **Config Class**: `Mistral3Config` (multimodal wrapper) with `text_config` nested inside
- **Text Decoder Path**: `model.model.language_model`
- **Vocab Size**: 131,072 tokens (requires chunked LM head)
- **Default Dtype**: bfloat16

## Implementation Status

### Code Changes
1. **Adapter**: `hf_adapters/hf_ministral3.py`- Working implementation using standard GQA
2. **Auto-loader**: `auto_spyre_model.py` - Registered `Mistral3Config` and `Ministral3Config`
3. **Test Registry**: `tests/model_registry.py` - Added with `dtype: bfloat16`
4. **Block Tests**: `tests/test_block_cpu_vs_spyre.py` - Config maker and special model loading
5. **Test Helper**: `tests/_helpers.py` - Special handling for `Mistral3ForConditionalGeneration`
6. **Documentation**: `ARCHITECTURE.md`, `README.md` - Updated

### Passing Tests
1. **Block CPU vs Spyre** (`test_block_cpu_vs_spyre.py ministral3_14b`):
   - [Pass] Prefill mode: All 40 layers pass (max diff ~0.4, within fp16 precision)
   - [Fail] Decode mode: Known torch-spyre limitation (batchmatmul restickify error for seq_len=1)

2. **E2E Smoke Test** (`test_e2e_smoke_spyre.py ministral3_14b`):
   - [Pass] Model loads successfully (115.3s)
   - [Pass] Generates 5 tokens on Spyre (87.3s)

3. **Auto-loader CPU Accuracy** (`test_auto_loader[ministral3_14b]`):
   -  Pass

## Issue

#### Incorrect Generated Output
**Summary**: Although model loading, compilation, and block-level validation complete successfully, generated text is consistently incoherent..

<img width="1509" height="576" alt="Image" src="https://github.com/user-attachments/assets/7a1d7834-843a-448a-91af-4a389be7371b" />

**Observed in**:
- E2E smoke test output: 'rollers πλη competingIDSeca'
- Manual generation test: garbled non-English characters

**Possible Causes**:
- Potential numerical drift accumulation across autoregressive steps

## Technical Details

### Multimodal Config Structure
```python
Mistral3Config
├── text_config (Ministral3Config)
│   ├── hidden_size: 5120
│   ├── num_attention_heads: 40
│   ├── num_key_value_heads: 8
│   ├── head_dim: 128
│   └── vocab_size: 131072
└── vision_config (...)
```

### Adapter Approach
- Uses standard GQA compilation (`make_standard_gqa_block`)
- Chunked LM head (8 chunks) for large vocabulary
- Config attribute propagation from `text_config` to top level
- RMSNorm patching (tries `Mistral3RMSNorm`, falls back to `MistralRMSNorm`)

## Files Modified
1. `hf_adapters/hf_ministral3.py` - Adapter implementation
2. `hf_adapters/auto_spyre_model.py` - Config mappings
3. `tests/model_registry.py` - Test registry entry
4. `tests/test_block_cpu_vs_spyre.py` - Block test support
5. `tests/_helpers.py` - Special model loading
6. `ARCHITECTURE.md` - Documentation
7. `README.md` - Model listing

## Attachments
<img width="1506" height="368" alt="Image" src="https://github.com/user-attachments/assets/e0d265a8-4de0-4f99-8cee-63635731c476" />
<img width="987" height="101" alt="Image" src="https://github.com/user-attachments/assets/59d0f569-cdc5-4e49-a81f-d129b2b0456c" />
<img width="1172" height="902" alt="Image" src="https://github.com/user-attachments/assets/82ca17cc-94e6-4140-92d3-f01a61ca96bf" />

**Comment by Mohini-Shrivastava** [comment]
PR for Mistral3 model : https://github.com/torch-spyre/hf-adapters/pull/83

**Comment by Mohini-Shrivastava** [comment]
Hi @arielge , 
Looks like a simple inference on CPU is generating absurd results with Mistral3

```
import torch
from huggingface_hub import hf_hub_download
from PIL import Image
from transformers import AutoProcessor

SAMPLE_IMAGE = {
"repo_id": "huggingface/documentation-images",
"filename": "pipeline-cat-chonk.jpeg",
"repo_type": "dataset",
}
model_path = "mistralai/Mistral-Small-3.1-24B-Instruct-2503"
path = hf_hub_download(**SAMPLE_IMAGE)
image = Image.open(path).convert("RGB")
processor = AutoProcessor.from_pretrained(model_path,fix_mistral_regex=True)
prompt = "Briefly describe this image"
conv = [
{
"role": "user",
"content": [
{"type": "image", "image": image},
{"type": "text", "text": prompt},
],
}
]
batch = processor.apply_chat_template(
conv,
add_generation_prompt=True,
tokenize=True,
return_dict=True,
return_tensors="pt",
)
from transformers import AutoModelForImageTextToText
max_new_tokens = 16
ref_model = AutoModelForImageTextToText.from_pretrained(
model_path, device_map="cpu"
).eval()
prompt_len = batch["input_ids"].shape[1]

with [torch.no](http://torch.no/)_grad():
gen = ref_model.generate(
**batch,
max_new_tokens=max_new_tokens,
do_sample=False,
use_cache=True,
)
text = processor.tokenizer.decode(gen[0, prompt_len:], skip_special_tokens=True)
print(text)
```

Result:
`"The image depicts a Pallas's cat, also known as the manul."`

cc: @BenjSz

**Comment by BenjSz** [comment]
Do we keep this one open? 
@arielge ? @Mohini-Shrivastava ?

**Comment by Mohini-Shrivastava** [comment]
@BenjSz, IMO we can close this.
cc: @arielge
